### ISSUE TITLE
Cilium BPF LB maps may get out of sync with k8s

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#7930]
Cilium version 1.4.3

Observed that traffic was being loadbalanced to a particular IP of an endpoint that is no longer present in the cluster. At the destination node, `cilium monitor -t drop` shows messages like this:

```
CPU 07: MARK 0x1ba71ba7 FROM 0 DROP: 74 bytes, reason Not a local target address
```

After running bugtool, I found that the destination IP is in some LB maps, specifically in the following sections of the bugtool:
- `cilium-debuginfo-xxx.md` - Under `Service list`
- `cilium-bpf-lb-list.md`

However, it is not present in the following sections:
- `cilium-debuginfo-xxx.md` - Under `k8s-service-cache`
- `iptables-save.md`

This warning was also observed in the cilium logs:
```
level=error msg="Error while inserting service in LB map" error="unable to update service <redacted>:80 with the value <redacted>:80 (144): Unable to update element for map with file descriptor 13: argument list too long" k8sNamespace=kube-ingress k8sSvcName=ingress-nginx subsys=daemon
````

**Comment by brb** [comment]
Do we have cilium-agent logs from the source node?

At the high level, the service update consists of the following steps:

goroutine A:
 
1. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/daemon/k8s_watcher.go#L382
2. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/daemon/k8s_watcher.go#L794
3. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/daemon/k8s_watcher.go#L790
4. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/pkg/k8s/service_cache.go#L132-L137

goroutine B:

1. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/daemon/k8s_watcher.go#L731-L732
2. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/daemon/k8s_watcher.go#L924
3. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/daemon/loadbalancer.go#L124
4. https://github.com/cilium/cilium/blob/31d8e021989b26dcca85028704f0192d75dfc5e4/daemon/loadbalancer.go#L129

Considering, that `iptables-save.md` and `k8s-service-cache` do not contain the entry, but `Service list` (updated in B.4.) and `cilium-bpf-lb-list.md` (updated in B.3.) do, I expect that the B.3. step has failed OR there was a race condition - TCP SYN sent to the endpoint while the daemon was removing the endpoint from the service, but that requires the svc update event to be send after the pod has been removed.

Having the logs would help to know what exactly has failed.

**Comment by brb** [comment]
Just checked, that some service maps are full (mostly due to duplicate endpoint entries, resolved in v1.5), e.g. `cilium-6npj5` contains 65544 entries in the `cilium_lb4_services` map, while the map size is 65536 (which is less than the entries count! UPDATE: @borkmann gave a hint that it might be due to updates happening while dumping the BPF map). So, updating the map should have failed in the B.3. step. Thus, we have the out of sync.

**Comment by felipejfc** [comment]
@brb do you have any hypothesis on how could the map grow so large? It is clearly leaking pod ips inside service definitions when we do ```cilium bpf lb list```, there are some ips of pods that doesn't exist in the cluster anymore...

**Comment by brb** [comment]
@felipejfc Can you please send me cilium-agent logs? I'm `martynas` on the Cilium slack.

**Comment by felipejfc** [comment]
just sent them @brb !

**Comment by brb** [comment]
Thanks for the logs.

Yes, the BPF maps are full, thus the out-of-sync. For example there are many errors like this one:

```
1556881042990,level=error msg="Error while inserting service in LB map" error="unable to update service 100.67.198.177:80 with the value 100.96.82.196:8080 (329): Unable to update element for map with file descriptor 10: argument list too long" k8sNamespace=push-api k8sSvcName=push-api subsys=daemon
```

And this is because of the duplicate entries problem resolved in v1.5 by #7540:

```
> wc -l cilium-bpf-lb-list.md
65544 cilium-bpf-lb-list.md
# considering that no enpoint appears in two different services
> cat cilium-bpf-lb-list.md | sort | uniq | wc -l
1671
```
> there are some ips of pods that doesn't exist in the cluster anymore

@felipejfc Could you point to any?

**Comment by brb** [comment]
This has been resolved in v1.5.
